sklearn/preprocessing/data.py

```diff
@@ -2567,3 +2577,299 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+    values per feature and transform the data to a binary one-hot encoding.
+
+    >>> from sklearn.preprocessing import CategoricalEncoder
+    >>> enc = CategoricalEncoder(handle_unknown='ignore')
+    >>> X = [['Male', 1], ['Female', 3], ['Female', 2]]
+    >>> enc.fit(X)
+    ... # doctest: +ELLIPSIS
+    CategoricalEncoder(categories='auto', dtype=<... 'numpy.float64'>,
+              encoding='onehot', handle_unknown='ignore')
+    >>> enc.categories_
+    [array(['Female', 'Male'], dtype=object), array([1, 2, 3], dtype=object)]
+    >>> enc.transform([['Female', 1], ['Male', 4]]).toarray()
+    array([[ 1.,  0.,  1.,  0.,  0.],
+           [ 0.,  1.,  0.,  0.,  0.]])
+    >>> enc.inverse_transform([[0, 1, 1, 0, 0], [0, 0, 0, 1, 0]])
+    array([['Male', 1],
+           [None, 2]], dtype=object)
+
+    See also
+    --------
+    sklearn.preprocessing.OneHotEncoder : performs a one-hot encoding of
+      integer ordinal features. The ``OneHotEncoder assumes`` that input
+      features take on values in the range ``[0, max(feature)]`` instead of
+      using the unique values.
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
+      dictionary items (also handles string-valued features).
+    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
+      encoding of dictionary items or strings.
+    """
+
+    def __init__(self, encoding='onehot', categories='auto', dtype=np.float64,
+                 handle_unknown='error'):
+        self.encoding = encoding
+        self.categories = categories
+        self.dtype = dtype
+        self.handle_unknown = handle_unknown
+
+    def fit(self, X, y=None):
+        """Fit the CategoricalEncoder to X.
+
+        Parameters
+        ----------
+        X : array-like, shape [n_samples, n_features]
+            The data to determine the categories of each feature.
+
+        Returns
+        -------
+        self
+
+        """
+        if self.encoding not in ['onehot', 'onehot-dense', 'ordinal']:
+            template = ("encoding should be either 'onehot', 'onehot-dense' "
+                        "or 'ordinal', got %s")
+            raise ValueError(template % self.handle_unknown)
+
+        if self.handle_unknown not in ['error', 'ignore']:
+            template = ("handle_unknown should be either 'error' or "
+                        "'ignore', got %s")
+            raise ValueError(template % self.handle_unknown)
+
+        if self.encoding == 'ordinal' and self.handle_unknown == 'ignore':
+            raise ValueError("handle_unknown='ignore' is not supported for"
+                             " encoding='ordinal'")
+
+        if self.categories != 'auto':
+            for cats in self.categories:
+                if not np.all(np.sort(cats) == np.array(cats)):
+                    raise ValueError("Unsorted categories are not yet "
+                                     "supported")
+
+        X_temp = check_array(X, dtype=None)
```

Comment:
It seems a bit strange that if X is a pandas DataFrame and we're accessing it column-wise we first need to stick it into a contiguous numpy array. Shrug.

Reply:
That is completely true... 
Would it be acceptable to make that a special case? Either here, or extend `check_array` with something like a `allow_dataframe` keyword ?

Reply:
It's not that simple as you still need to recognise here that it's a dataframe to access with iloc rather than direct indexing. So you need to bypass check_array if it's a DF, and then after getting n_samples and n_features, set `X = X.iloc`

Reply:
Ah, yes, indeed not that simple. Certainly because `iloc` does not support boolean masking, which is needed here (although a `np.asarray` call after accessing the column would be enough to solve that). 
But still doable *if* we want this (but can also in a follow up PR, as that might stir some more discussion)

Reply:
No, I think boolean masking on the series extracted by iloc is fine...

Reply:
Hmm, I was convinced it was a problem. I even added a workaround for it in the column transformers PR (https://github.com/scikit-learn/scikit-learn/pull/9012/files/#diff-4f9e3dd0dbe9e8dffb16e987cc43f38bR386), but can't reproduce it now ... Anyhow, that's another discussion :-)